### PR TITLE
feat: add possibility to export projects as JSON

### PR DESCRIPTION
**Pull request by ccailly**
Added the ability to export gantt view data in JSON format.

image

### CODE REVIEW

File: src/ProjectsExport.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+namespace GlpiPlugin\Gantt;
+
+class ProjectsExport
```
Comment (AdrienClairembault) [inline src/ProjectsExport.php:38]:
```suggestion
final class ProjectsExport
```

File: front/projectsexport.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+use GlpiPlugin\Gantt\ProjectsExport;
+
+include('../../../inc/includes.php');
+
+$project_id = $_GET['project_id'] ?? -1;
+
+// Check right
+Session::checkRight(Project::$rightname, Project::READMY);
+
```
Comment (AdrienClairembault) [inline front/projectsexport.php:44]: I guess we should also call `$project->canViewItem()` to make sure an user can't export someone else project that they are not allowed to see.

File: src/ProjectsExport.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    public function json(): string
+    {
+        header_remove('Content-Type');
+        header_remove('Content-Disposition');
+
+        header('Content-Type: application/json');
+        header('Content-Disposition: attachment; filename="gantt-data.json"');
+
+        return json_encode($this->loadProjectsData());
```
Comment (AdrienClairembault) [inline src/ProjectsExport.php:116]: We usually try to define the headers in the `front` file.
Can you move the header to the front file and just keep the `return json_encode` part here ?

`header_remove` shouldn't be needed too, did you encounter some issues that forced you to use it ?

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]

File: front/projectsexport.php (hunk 2 above)
Comment (ccailly) [reply front/projectsexport.php:44]: The `getItemsForProject` method already performs the necessary rights checks.

https://github.com/pluginsGLPI/gantt/blob/36978f33ad541b831bd081a34a1fbe8648e5ca7b/src/ProjectsExport.php#L49-L50

https://github.com/pluginsGLPI/gantt/blob/5701bb66f72d49241f6631d11a9909e1192493cc/src/DataFactory.php#L47-L71

File: src/ProjectsExport.php (hunk 3 above)
Comment (ccailly) [reply src/ProjectsExport.php:116]: I've added `header_remove` as a preventive measure to ensure that the headers are good, but in the end they're useless as I'm writing over them again. I've had no problems without them.

File: front/projectsexport.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+include('../../../inc/includes.php');
+
+$project_id = $_GET['project_id'] ?? -1;
+
+// Check right
+Session::checkRightsOr(Project::$rightname, [Project::READALL, Project::READMY]);
+
+header('Content-Type: application/json');
+header('Content-Disposition: attachment; filename="gantt-data.json"');
```
Comment (Rom1-B) [inline front/projectsexport.php:46]: When exporting from a project, I think the project name should be added to the file name.

**Review by Rom1-B — CHANGES_REQUESTED** [review]

**Review by AdrienClairembault — APPROVED** [review]

**Pull request merged**